modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -290,37 +297,56 @@ public CompletableFuture<Void> finish(
-
-                                TxStateMeta finalTxStateMeta = coordinatorFinalTxStateMeta(commit, commitTimestamp);
-
-                                finishingStateMeta.txFinishFuture().complete(finalTxStateMeta);
-
-                                return finalTxStateMeta;
-                            })
-                    );
+            // In case of commit it's required to check whether current primaries are still the same that were enlisted and whether
+            // given primaries are not expired or, in other words, whether commitTimestamp is less or equal to the enlisted primaries
+            // expiration timestamps.
+            CompletableFuture<Void> verificationFuture =
+                    commit ? verifyCommitTimestamp(enlistedGroups, commitTimestamp) : completedFuture(null);
+
+            return verificationFuture.handle(
+                    (unused, throwable) -> {
+                        Collection<ReplicationGroupId> replicationGroupIds = new HashSet<>(enlistedGroups.keySet());
+
+                        boolean verifiedCommit = throwable == null && commit;
+
+                        TxFinishReplicaRequest req = FACTORY.txFinishReplicaRequest()
+                                .txId(txId)
+                                .timestampLong(clock.nowLong())
+                                .groupId(commitPartition)
+                                .groups(replicationGroupIds)
+                                // In case of verification future failure transaction will be rolled back.
+                                .commit(verifiedCommit)
+                                .commitTimestampLong(hybridTimestampToLong(commitTimestamp))
+                                .term(term)
+                                .build();
+
+                        return replicaService.invoke(recipientNode, req).thenRun(
+                                () -> {
+                                    updateTxMeta(txId, old -> {
+                                        if (isFinalState(old.txState())) {
+                                            finishingStateMeta.txFinishFuture().complete(old);
+
+                                            return old;
+                                        }
+
+                                        assert old instanceof TxStateMetaFinishing;
+
+                                        TxStateMeta finalTxStateMeta = coordinatorFinalTxStateMeta(verifiedCommit, commitTimestamp);
+
+                                        finishingStateMeta.txFinishFuture().complete(finalTxStateMeta);
+
+                                        return finalTxStateMeta;
+                                    });
+
+                                    if (verifiedCommit) {
+                                        observableTimestampTracker.update(commitTimestamp);
+                                    }
+                                });
+                    })
+                    .thenCompose(Function.identity())
```

Comment:
It this invokation required?

Reply:
Don't think so. Removed.

Reply:
My fault. It's required. Otherwise exceptions from original future won't be thrown. Lot's of tests including for example org.apache.ignite.internal.table.TxAbstractTest#testTransactionAlreadyCommitted will fail. Restored.